lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -2623,6 +2633,43 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
 
     walletId = getApiT apiWalletId
 
+    parseReferenceScript
+        :: ApiConstructTransactionData n
+        -> (SlotNo, SlotNo)
+        -> Either ErrConstructTx (Maybe (Script Cosigner))
+    parseReferenceScript tx validity = do
+        let mbRefScript = tx ^. #referencePolicyScriptTemplate
+        for mbRefScript $ \(ApiT refScript) -> do
+            guardWrongScriptTemplate refScript
+            guardOutsideValidityInterval validity refScript
+            Right refScript
+      where
+        guardWrongScriptTemplate
+            :: Script Cosigner -> Either ErrConstructTx ()
+        guardWrongScriptTemplate apiScript =
+            when (wrongMintingTemplate apiScript)
+                $ Left ErrConstructTxWrongMintingBurningTemplate
+          where
+            wrongMintingTemplate script =
+                isLeft (validateScriptOfTemplate RecommendedValidation script)
+                || countCosigners script /= (1 :: Int)
+                || existsNonZeroCosigner script
+            countCosigners = foldScript (const (+ 1)) 0
+            existsNonZeroCosigner =
+                foldScript (\cosigner a -> a || cosigner /= Cosigner 0) False
+
+        guardOutsideValidityInterval
+            :: (SlotNo, SlotNo)
+            -> Script Cosigner
+            -> Either ErrConstructTx ()
+        guardOutsideValidityInterval (before, hereafter) script =
+            when notWithinValidityInterval $
+                Left ErrConstructTxValidityIntervalNotWithinScriptTimelock
+          where
+            notWithinValidityInterval =
+                not $ withinSlotInterval before hereafter $
+                    scriptSlotIntervals script
```

Comment:
Same question as above, with the following question added: As the script is supposed to be used as a reference input, it would be ok for the script to have a much larger validity interval (`scriptSlotIntervals script`) than the transaction which is used to put the script on chain (`(before, herafter)`)? I.e. I believe that `guardOutsideValidityInterval` should not be present. 🤔

Reply:
there is ledger rule that requires script timelocks determined by (hereafter, before) to encapsulate validity interval. If not the case ledger will not accept this. Ask Jared why is that way ;-)

Reply:
please see the following integration test https://github.com/cardano-foundation/cardano-wallet/blob/master/lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L3600 . Generally speaking we need script's timelock interval be in validity interval. If it is not the case, as in the test above node will end up with error upon submitting. This is checked somewhere in ledger code (not in UTXOW rules specified unfortunately)

Reply:
you are right! when script is published to be used later it does not make any sense to check against validity as we do not know when it is going to be used (we know only that later). Moreover we cannot use the published script in the same transaction.

Thanks @HeinrichApfelmus for raising this very subtle thing! Good catch!

I have removed this validation!

Reply:
🤓 Thanks!